### PR TITLE
Improve L-BFGS with Moré-Thuente line search and numerical safeguards

### LINKED ISSUE #16236: Wrong definition of interval length in zoom linesearch
### Description

My issue is in the definition of the interval length in the zoom linesearch in [jax/_src/scipy/optimize/line_search.py](https://github.com/google/jax/blob/main/jax/_src/scipy/optimize/line_search.py) on [line 113](https://github.com/google/jax/blob/cfabad58867c2f3fa46824a5462e4db7dd9f229f/jax/_src/scipy/optimize/line_search.py#L113). 

The interpolation method aims at finding a point `a_j` lying in the interval with boundary points `a_lo`, `a_hi`, that is `[a_lo, a_hi]` or `[a_hi, a_lo]` depending on their order, as stated in Algorithm 3.6 of [Wright and Nocedal, 'Numerical Optimization', 1999](https://pages.mtu.edu/~struther/Courses/OLD/Sp2013/5630/Jorge_Nocedal_Numerical_optimization_267490.pdf).

A priori nothing prevents the algorithm as stated by Wright and Nocedal to have `a_hi < a_lo` since high and low refer to function values not left and right sides of the interval (in fact it can even be initialized with `a_hi < a_lo`, see third case in Algorithm 3.5 of Wright and Nocedal). 

Yet in the current implementation the algorithm will fail if `dalpha = state.a_hi - state.a_lo < 0` as per  [line 122](https://github.com/google/jax/blob/cfabad58867c2f3fa46824a5462e4db7dd9f229f/jax/_src/scipy/optimize/line_search.py#L122). Moreover the checks on [line 127](https://github.com/google/jax/blob/cfabad58867c2f3fa46824a5462e4db7dd9f229f/jax/_src/scipy/optimize/line_search.py#L127) and [line 129](https://github.com/google/jax/blob/cfabad58867c2f3fa46824a5462e4db7dd9f229f/jax/_src/scipy/optimize/line_search.py#L129) are a priori wrong since `cchk = delta1*dalpha = 0.1*(state.a_hi - state.a_lo)` can be negative meaning that we may select a point satisfying `(a_j_cubic > a + cchk) & (a_j_cubic < b - cchk)` while having `a_j_cubic < a` or `a_j_cubic > b`, that is we may select a point outside the desired interval. 

Note that the second issue (wrong checks for cubic and quadratic) seems to be also present in Scipy's original code [line 544](https://github.com/scipy/scipy/blob/c1ed5ece8ffbf05356a22a8106affcd11bd3aee0/scipy/optimize/_linesearch.py#L544), [line 562](https://github.com/scipy/scipy/blob/c1ed5ece8ffbf05356a22a8106affcd11bd3aee0/scipy/optimize/_linesearch.py#L562) and [line 564](https://github.com/scipy/scipy/blob/c1ed5ece8ffbf05356a22a8106affcd11bd3aee0/scipy/optimize/_linesearch.py#L564).

This issue can simply be fixed by defining `dalpha = jnp.abs(state.a_hi - state.a_lo)` on line 113.

### What jax/jaxlib version are you using?

_No response_

### Which accelerator(s) are you using?

_No response_

### Additional system info

_No response_

### NVIDIA GPU info

_No response_

### LINKED ISSUE #36793: L-BFGS: improve robustness with Moré-Thuente line search and numerical safeguards
### Description

The current L-BFGS implementation (`jax._src.scipy.optimize._lbfgs`) has several robustness issues compared to scipy's `L-BFGS-B` that cause it to fail on non-trivial optimization problems. I ran into this while fitting a QGARCH(1,1) financial model — L-BFGS converges on only ~15-30% of series, while scipy's L-BFGS-B converges on 100%.

I've been digging into the root causes and I think there are three main things going on:

**1. Missing `old_old_fval` in line search**

The BFGS implementation already tracks `old_old_fval` and passes it to the line search for adaptive initial step size estimation (see `bfgs.py` line 121, 138, 169). L-BFGS never does this — it always starts the line search at `alpha=1.0`. This is a significant oversight since the same line search code already supports it.

**2. The Wolfe line search is less robust than Moré-Thuente**

Scipy's L-BFGS-B uses the Moré-Thuente (DCSRCH) line search from MINPACK-2, which has a two-stage approach (modified function then direct) with cubic/quadratic/bisection safeguards. The current JAX implementation uses Algorithm 3.5/3.6 from Wright & Nocedal, which is simpler but fails more often on harder problems.

I've ported the Moré-Thuente algorithm to JAX using `lax.switch` for the four dcstep cases (so only the selected case executes at runtime) and `lax.cond` for the modified-vs-direct function path. The key subtlety was handling negative discriminants in the cubic interpolation — the original code uses `sqrt(discriminant)` which produces NaN when the discriminant is negative. The fix is to fall back to bisection in those cases.

**3. No numerical safeguards for rho_k and gamma**

When `y·s == 0`, `rho_k = 1/0 = inf` which poisons the history. When `y·y == 0`, `gamma = 0/0 = NaN`. Scipy handles both cases (e.g. `rhok = 1000.0` when `y·s == 0`). Adding similar guards prevents NaN cascades.

### Results

I have a working implementation that matches scipy's accuracy on my benchmark (200 QGARCH series, T=800):

- 199/200 series find the same minimum as scipy L-BFGS-B (NLL within 0.001)
- 2.0ms/series vs scipy's 2.6ms/series
- The 1 remaining series actually finds a *better* minimum than scipy

Also fixes the `dalpha` sign bug reported in #16236 as part of the line search changes.

### Proposed changes

- New file `dcsrch.py`: Moré-Thuente line search for JAX
- Modify `_lbfgs.py`: use new line search, track `old_old_fval`, add numerical safeguards
- Fix `line_search.py`: `dalpha` sign bug (#16236) and negative initial step guard

I have the implementation ready on my fork and can open a PR if there's interest.

### Future work

A bounded variant (L-BFGS-B) with gradient projection and Cauchy point computation would also be valuable but is a separate, larger effort.

### System info

jax 0.9.2, Python 3.11, macOS (Apple Silicon)

### PR DESCRIPTION
**Pull request by jjmorenog620**
Fixes #36793, fixes #16236, related to #19078

### Summary

The L-BFGS optimizer has several robustness issues that cause it to fail on non-trivial optimization problems. This PR replaces the Wright-Nocedal Wolfe line search with the Moré-Thuente (DCSRCH) algorithm from MINPACK-2, adds adaptive initial step size estimation, and introduces numerical safeguards, bringing the implementation to parity with scipy's L-BFGS-B.

**Moré-Thuente line search** (`dcsrch.py`): A new line search implementation using the two-stage DCSRCH algorithm (modified function then direct minimization). The four dcstep cases are dispatched via `lax.switch` so only the selected case executes at runtime. A discriminant guard on all cases falls back to bisection when cubic interpolation is invalid, which was the key fix for convergence on hard problems.

**Adaptive initial step size**: L-BFGS now tracks `old_old_fval` across iterations and uses it to compute the initial line search step size, matching what BFGS already does (`bfgs.py` lines 67, 121, 138, 169). Previously L-BFGS always started the line search at alpha=1.0.

**Numerical safeguards**: `rho_k` is set to 1000.0 when `y·s == 0` (matching scipy's convention, prevents inf from poisoning the history). `gamma` preserves the previous value when `y·y == 0` (prevents NaN in the two-loop recursion). On line search failure, the optimizer takes a small gradient descent step instead of stopping immediately.

**Line search fixes** (`line_search.py`): The `dalpha` computation in `_zoom` now uses `jnp.abs` to handle reversed brackets correctly (#16236). Negative initial step candidates are rejected and defaulted to 1.0, matching scipy's guard.

### Results

Tested on a QGARCH(1,1) maximum likelihood estimation problem (200 Monte Carlo series, T=800, 4 parameters) and standard test functions (Rosenbrock, Beale, Himmelblau):

|  | Before | After | scipy L-BFGS-B |
|--|--------|-------|----------------|
| Series matching scipy (NLL within 0.001) | ~30/200 | 200/200 | baseline |
| Parameter mean difference vs scipy | 0.12 | 0.000006 | baseline |
| Speed per series | 0.6ms | 1.3ms | 2.6ms |

All 200 series now converge to the same minimum as scipy L-BFGS-B, with parameter estimates matching to 6 decimal places, at twice the speed. Standard test functions converge to the correct minimum.

### Future work

A bounded variant (L-BFGS-B) with gradient projection and Cauchy point computation would be a natural next step but is a separate, larger effort.

### CODE REVIEW

**Review by jjmorenog620 — COMMENTED** [review]
Updated — instead of the hardcoded `rho_k = 1000.0`, the history update is now skipped entirely when `rho_k` is invalid (`rho_k_inv == 0` or non-finite). The history vectors and scalars keep their previous values, so the two-loop recursion continues with clean data. This matches the pattern BFGS already uses (`bfgs.py` line 159: `H_kp1 = jnp.where(jnp.isfinite(rho_k), H_kp1, state.H_k)`).

File: jax/_src/scipy/optimize/_lbfgs.py [hunk 1]
```diff
@@ -161,25 +169,40 @@ def body_fun(state: LBFGSResults):
...
+      ls_results.a_k,
     )
 
     # evaluate at next iterate
-    s_k = jnp.asarray(ls_results.a_k).astype(p_k.dtype) * p_k
+    s_k = jnp.asarray(safe_alpha).astype(p_k.dtype) * p_k
     x_kp1 = state.x_k + s_k
-    f_kp1 = ls_results.f_k
-    g_kp1 = ls_results.g_k
+    f_kp1, g_kp1 = api.value_and_grad(fun)(x_kp1)
```
[the remark this answers is not in the document]
Comment (jjmorenog620) [reply jax/_src/scipy/optimize/_lbfgs.py]: Fixed — now using `lax.cond` to reuse line search results and only re-evaluate on fallback.

File: jax/_src/scipy/optimize/_lbfgs.py [hunk 2]
```diff
@@ -161,25 +169,40 @@ def body_fun(state: LBFGSResults):
...
+    )
+    y_dot_y = jnp.real(_dot(jnp.conj(y_k), y_k))
+    gamma = jnp.where(y_dot_y > 0, rho_k_inv / y_dot_y, state.gamma)
 
     # replacements for next iteration
+    ls_ok = ~ls_results.failed & jnp.isfinite(ls_results.a_k) & (ls_results.a_k > 0)
     status = jnp.array(0)
-    status = jnp.where(state.f_k - f_kp1 < ftol, 4, status)
+    # Only count insufficient progress when line search actually succeeded.
+    status = jnp.where(ls_ok & (state.f_k - f_kp1 < ftol), 4, status)
```
[the remark this answers is not in the document]
Comment (jjmorenog620) [reply jax/_src/scipy/optimize/_lbfgs.py]: Fixed — restored `ftol` check for all steps.

File: jax/_src/scipy/optimize/_lbfgs.py [hunk 3]
```diff
@@ -153,6 +154,13 @@ def body_fun(state: LBFGSResults):
     # find search direction
     p_k = _two_loop_recursion(state)
 
+    # Compute initial step size from old_old_fval (same as BFGS).
+    dphi0 = jnp.real(_dot(state.g_k, p_k))
+    alpha0 = jnp.where(
+      dphi0 != 0,
+      jnp.clip(1.01 * 2 * (state.f_k - state.old_old_fval) / dphi0, 1e-10, 1.0),
```
[the remark this answers is not in the document]
Comment (jjmorenog620) [reply jax/_src/scipy/optimize/_lbfgs.py]: Fixed — now defaults to 1.0 when `f_k >= old_old_fval`.

File: jax/_src/scipy/optimize/dcsrch.py [hunk 4]
```diff
@@ -0,0 +1,385 @@
...
+      stage=upd_stage,
+      finit=state.finit, ginit=state.ginit, gtest=state.gtest,
+      width=new_width, width1=new_width1,
+      stmin=new_stmin, stmax=new_stmax,
+      done=done, failed=failed, nfev=nfev,
+    )
+
+  final = lax.while_loop(cond, body, init_state)
+
+  _, _, g_star = full_eval(final.stp)
```
[the remark this answers is not in the document]
Comment (jjmorenog620) [reply jax/_src/scipy/optimize/dcsrch.py]: Fixed — gradient vector is now stored in `_DCSRCHState.g_full`, eliminating the extra evaluation.

File: jax/_src/scipy/optimize/dcsrch.py [hunk 5]
```diff
@@ -0,0 +1,385 @@
...
+
+  return _LineSearchResults(
+    failed=final.failed,
+    nit=final.nfev,
+    nfev=final.nfev + 1,
+    ngev=final.nfev + 1,
```
[the remark this answers is not in the document]
Comment (jjmorenog620) [reply jax/_src/scipy/optimize/dcsrch.py]: Fixed — `init_nfev` now accounts for the initial evaluation when `old_fval` or `gfk` are not provided.

File: jax/_src/scipy/optimize/_lbfgs.py [hunk 6]
```diff
@@ -161,25 +169,40 @@ def body_fun(state: LBFGSResults):
...
     x_kp1 = state.x_k + s_k
-    f_kp1 = ls_results.f_k
-    g_kp1 = ls_results.g_k
+    f_kp1, g_kp1 = api.value_and_grad(fun)(x_kp1)
     y_k = g_kp1 - state.g_k
     rho_k_inv = jnp.real(_dot(y_k, s_k))
-    rho_k = jnp.reciprocal(rho_k_inv).astype(y_k.dtype)
-    gamma = rho_k_inv / jnp.real(_dot(jnp.conj(y_k), y_k))
+    rho_k = jnp.where(
+      rho_k_inv == 0., jnp.array(1000., dtype=y_k.dtype),
```
[the remark this answers is not in the document]
Comment (jjmorenog620) [reply jax/_src/scipy/optimize/_lbfgs.py]: Updated — instead of the hardcoded `rho_k = 1000.0`, we now skip the history update entirely when `rho_k` is invalid (`rho_k_inv == 0` or non-finite). The history vectors and scalars keep their previous values, so the two-loop recursion continues with clean data. This matches the pattern BFGS already uses (`bfgs.py` line 159: `H_kp1 = jnp.where(jnp.isfinite(rho_k), H_kp1, state.H_k)`).
